app/scripts/lib/experiments/base.js

```diff
@@ -0,0 +1,181 @@
+
+    /**
+     * Initialize the experiment
+     *
+     * @returns {boolean}
+     *  Returns 'true' when the experiment successfully initialized. Otherwise 'false'.
+     */
+    initialize: function (name, options) {
+      this._initialized = false;
+      // all experiments require these options
+      if (! name || ! options || ! options.window || ! options.metrics || ! options.able || ! options.user || ! options.notifications) {
+        return false;
+      }
+
+      this.window = options.window;
+      this.able = options.able;
+      this.metrics = options.metrics;
+      this.user = options.user;
+      this.extraAbleOptions = options.extraAbleOptions || {};
+      this._notifications = options.notifications;
+
+      var abData = {
+        isMetricsEnabledValue: this.metrics.isCollectionEnabled(),
+        uniqueUserId: this.user.get('uniqueUserId'),
+        // the window parameter will override any ab testing features
+        forceExperimentGroup: Url.searchParam(FORCE_GROUP_TYPE, this.window.location.search)
+      };
+
+      abData = _.extend(abData, this.extraAbleOptions);
+
+      this._name = name;
+      this._groupType = this.able.choose(this._name, abData) || null;
+      if (! this._groupType) {
+        return false;
+      }
+
+      this._loggingNamespace = 'experiment.' + this._groupType + '.' + this._name + '.';
+      this._storageNamespace = 'experiment.' + this._name;
+      this._initialized = true;
+
+      this.saveState('enrolled');
+
+      this.delegateNotifications();
+
+      return true;
+    },
+
+    delegateNotifications: function (notifications) {
+      // based on delegateEvents from Backbone.View
+      if (! (notifications || (notifications = _.result(this, 'notifications')))) {
+        return false;
+      }
+
+      for (var notificationName in notifications) {
+        var method = notifications[notificationName];
+        if (_.isString(method)) {
+          method = this[notifications[notificationName]];
+        }
+
+        if (_.isFunction(method)) {
+          this._notifications.on(notificationName, method.bind(this));
+        }
+      }
+    },
+
+    /**
+     * Checks the experiment group
+     *
+     * @param groupType
+     * @returns {boolean}
+     */
+    isGroup: function (groupType) {
+      return !! (groupType && groupType === this._groupType);
+    },
+
+    /**
+     * Log an experiment event
+     *
+     * @param event
+     */
+    logEvent: function (event) {
+      if (this._initialized && event) {
+        this.metrics.logEvent(this._loggingNamespace + event);
+      }
+    },
+
+    /**
+     * Saves some 'state' value into experiment storage.
+     *
+     * State is used to keep track of what happened during the experiment.
+     *
+     * @param state
+     */
+    saveState: function (state) {
+      if (! state) {
+        return false;
+      }
+
+      var store = {};
+      try {
+        store = JSON.parse(storage.get(this._storageNamespace)) || {};
```

Comment:
Instead of directly using the singleton `storage` object and calling [Session.clear in the unit tests](https://github.com/mozilla/fxa-content-server/pull/2943/files#diff-ff73ef0efdebab77e9c7b793ae9b396cR57), it'd be easier to pass in an optional `storage` instance to initialize. If the instance is passed in, use it, if not, use the singleton instance. The tests would use a NullStorage instance.